### PR TITLE
Check if we have a proper fileinfo

### LINKED ISSUE #14390: OC\Files\Node\Node doesn't check it has a valid FileInfo object before working on it
### Problem
A method creates a new OC\Files\Node\Node, using a broken path
A method then calls OC\Files\Node\Node->getFileInfo()
```
public function getFileInfo() {
	if (!$this->fileInfo) {
		$this->fileInfo = $this->view->getFileInfo($this->path);
	}
	return $this->fileInfo;
}
```
which calls OC\Files\View->getFileInfo using the broken path
```
public function getFileInfo($path, $includeMountPoints = true) {
...
		$data = array();
		if (!Filesystem::isValidPath($path)) {
			return $data;
		}
...
```
which returns an empty array as a FileInfo object.

Now the Node object is going to call many different methods on an array and that will fail and break anything sitting on top of it.

Some examples of these methods
```
public function isMounted() {
	return $this->getFileInfo()->isMounted();
}
public function isShared() {
	return $this->getFileInfo()->isShared();
}
public function getMimeType() {
	return $this->getFileInfo()->getMimetype();
}
public function getMimePart() {
	return $this->getFileInfo()->getMimePart();
}
public function getType() {
	return $this->getFileInfo()->getType();
}
public function isEncrypted() {
	return $this->getFileInfo()->isEncrypted();
}
public function getMountPoint() {
	return $this->getFileInfo()->getMountPoint();
}
```

### Fix it in Node
* Check that we actually have a FileInfo object
* Throw an exception letting others know that things went sour

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes #14390

cc @PVince81 @oparoz

### CODE REVIEW

**Comment by PVince81** [comment]
Makes sense :+1:

File: lib/private/files/node/node.php [hunk 1]
```diff
@@ -60,11 +64,21 @@ public function __construct($root, $view, $path) {
 	/**
 	 * Returns the matching file info
 	 *
-	 * @return \OCP\Files\FileInfo
+	 * @return FileInfo
+	 * @throws InvalidPathException
```
Comment (DeepDiver1975) [inline lib/private/files/node/node.php]: PHPDoc on the interface has to be adjusted as well - behavior-wise a breaking change which has to be documented in the 8.1 features wiki - https://github.com/owncloud/core/wiki/ownCloud-8.1-Features#public-api

**Comment by DeepDiver1975** [comment]
Do unit tests cover this case? THX

**Comment by oparoz** [comment]
According to this message, this doesn't fix the problem
https://github.com/interfasys/galleryplus/issues/28#issuecomment-75768300
>Fatal error: Call to a member function getId() on a non-object in /var/www/owncloud/lib/private/files/node/node.php on line 156

**Comment by DeepDiver1975** [comment]
@icewind1991 please have a look - THX

**Comment by DeepDiver1975** [comment]
> Fatal error: Call to a member function getId() on a non-object in /var/www/owncloud/lib/private/files/node/node.php on line 156

That error message does not match up with the code in this pr

**Comment by oparoz** [comment]
@DeepDiver1975 - Probably because the patch was applied on `stable8`

**Comment by icewind1991** [comment]
Added unit test

**Comment by vader2014** [comment]
correct, i applied the change to stable8

**Comment by DeepDiver1975** [comment]
> Added unit test

:see_no_evil:

**Comment by icewind1991** [comment]
Actually push unit test....

File: lib/private/files/node/node.php (hunk 1 above)
Comment (MorrisJobke) [inline lib/private/files/node/node.php]: @icewind1991 This needs to be addressed ;)

**Comment by MorrisJobke** [comment]
Beside my comment above: :+1:

**Comment by DeepDiver1975** [comment]
:+1:

**Comment by DeepDiver1975** [comment]
we need a backport pr for stable8 @nickvergessen can you please help out here? THX

**Comment by DeepDiver1975** [comment]
@karlitschek backport requested because original issue was reported against oc8 - THX

**Comment by karlitschek** [comment]
backport is fine

**Comment by MorrisJobke** [comment]
Backport is in #14568

**Pull request merged**